File: src/learning_service_cpp/src/sum_client.cpp
/**
 * 需求：编写客户端 组织并发布两个整型数据作为请求 并正确接收解析响应
 * 步骤：
 *      1、包含头文件
 *      2、初始化 ROS2 客户端
 *      3、自定义节点类
 *          3-1、创建客户端
 *          3-2、等待连接服务器
 *          3-3、组织并发布请求
 *      4、创建节点对象指针 并处理响应
 *      5、释放资源
 */

#include "base_interfaces/srv/add_ints.hpp"
#include "rclcpp/rclcpp.hpp"

using namespace std::chrono_literals;
using base_interfaces::srv::AddInts;

class SumClient : public rclcpp::Node {
public:
    SumClient() : Node("sum_client_cpp") {
        RCLCPP_INFO(this->get_logger(), "客户端创建成功, 等待连接服务器...");
        client_ = this->create_client<AddInts>("sum");
    }

    // 等待连接服务器
    bool connect_server() {
        while (!client_->wait_for_service(1s)) {
            // 处理 ctrl + c 关闭异常
            if (!rclcpp::ok()) {
                RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "强制退出!");
                return false;
            }
            RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "服务器连接中...");
        }
        return true;
    }

    // 组织并发送请求
    rclcpp::Client<AddInts>::FutureAndRequestId send_request(int32_t num1, int32_t num2) {
        auto request = std::make_shared<AddInts::Request>();
        request->num1 = num1;
        request->num2 = num2;

        /*
            rclcpp::Client<base_interfaces::srv::AddInts>::FutureAndRequestId 
            async_send_request(std::shared_ptr<base_interfaces::srv::AddInts_Request> request)
        */
        return client_->async_send_request(request);
    }

private:
    rclcpp::Client<AddInts>::SharedPtr client_;
};

int main(int argc, char **argv) {
    // 处理请求数量异常情况
    if (argc != 3) {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "请提交两个整型数据!");
        return 1;
    }

    rclcpp::init(argc, argv);

    auto client_node = std::make_shared<SumClient>();
    if (client_node->connect_server()) {
        RCLCPP_INFO(rclcpp::get_logger("rclcpp"), "服务器连接成功!");
    } else {
        RCLCPP_ERROR(rclcpp::get_logger("rclcpp"), "服务器连接失败!");
        return 2;
    }

    // 发送请求 并处理响应结果
    auto future = client_node->send_request(atoi(argv[1]), atoi(argv[2]));
    if (rclcpp::spin_until_future_complete(client_node, future) == rclcpp::FutureReturnCode::SUCCESS) {
        // 请求处理正常
        RCLCPP_INFO(client_node->get_logger(), "请求处理正常!");
        RCLCPP_INFO(client_node->get_logger(), "响应结果: sum = %d", future.get()->sum);
    } else {
        // 请求处理异常
        RCLCPP_WARN(client_node->get_logger(), "请求处理异常!");
    }

    rclcpp::shutdown();

    return 0;
}

File: src/learning_service_py/learning_service_py/sum_client.py
"""
需求：编写客户端 组织并发送两个整型数据作为请求 并处理响应
步骤：
    1、导包
    2、初始化 ROS2 客户端
    3、自定义节点类
        3-1、创建客户端
        3-2、等待连接服务器
        3-3、编写发布请求函数
    4、创建节点对象 发送请求 并处理响应
    5、释放资源
"""

import sys
import rclpy
from rclpy.node import Node
from rclpy.logging import get_logger
from base_interfaces.srv import AddInts

class SumClient(Node):
    def __init__(self):
        super().__init__("sum_client_py")
        self.get_logger().info("客户端创建成功!(Python)")
        self.client_ = self.create_client(AddInts, "sum")

    def connect_server(self) -> bool:
        while not self.client_.wait_for_service(1.0):
            if not rclpy.ok():
                get_logger("rclpy").error("强制退出等待!")
                return False
            get_logger("rclpy").info("服务器连接中...")
        return True
    
    def send_request(self, num1: int, num2: int):
        request = AddInts.Request()
        request.num1 = num1
        request.num2 = num2
        return self.client_.call_async(request)

def main():
    if len(sys.argv) != 3:
        get_logger("rclpy").error("请提交两个整型数据!")
        return

    rclpy.init()

    sum_client = SumClient()

    connect_result = sum_client.connect_server()
    if connect_result:
        get_logger("rclpy").info("服务器连接成功!")
    else:
        get_logger("rclpy").error("服务器连接失败!")
    
    future = sum_client.send_request(int(sys.argv[1]), int(sys.argv[2]))
    rclpy.spin_until_future_complete(sum_client, future)

    try:
        response = future.result()
    except Exception as e:
        sum_client.get_logger().error(f"服务请求失败: {e}")
    else:
        sum_client.get_logger().info(f"响应结果为: sum = {response.sum}")

if __name__ == "__main__":
    main()